### PR TITLE
kokkos: add cmake_lang variant

### PR DESCRIPTION
**Pull request by rbberger**
kokkos in Spack currently relies on either direct compiler support for CUDA/HIP (`clang+nvptx`, `cce`) or makes use of the compiler wrappers `nvcc_wrapper` or `hipcc`.

Kokkos 3.6 added CMake language support via the `Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE` cmake option.

This PR adds a `cmake_lang` variant to enable this new mode. It is disabled by default, since any dependent package would have to add native support for CUDA and HIP CMake language. This change is a prerequisite for application/library developers to explore this option when using Kokkos.

### CODE REVIEW

**Comment by pranav-sivaraman** [comment]
It would be better to reuse the same functionality recently implemented in [cached_cmake.py](https://github.com/spack/spack/blob/develop/lib/spack/spack/build_systems/cached_cmake.py). That way, any future improvements there would eventually also land in the Kokkos Package. Let's say when a user wants to use Clang as the CUDA compiler instead of NVCC (Don't know if Kokkos support this). Currently, Spack doesn't really model this right now, but it's something I wish to see once compilers become dependencies.

**Comment by rbberger** [comment]
I agree having it centralized has its benefits, I'll take a closer look. Since I'd like to get rid of using hipcc altogether, CachedCMakePackage might need some modifications and might use a generic variant to enable/disable its use.

Btw, yes Kokkos already allows building with Clang as CUDA compiler, although it would be nice to detect nvptx support from the compiler spec.

**Review by masterleinad — COMMENTED** [review]
This will also need changes to `Kokkos` itself, see https://github.com/kokkos/kokkos/blob/e93b168babf9ce67e3a08268d109bcf9f654cdfd/CMakeLists.txt#L107-L109.

**Comment by cedricchevalier19** [comment]
@masterleinad, thank you for finding this issue.
I think we have to check if this `Kokkos` code is still relevant. https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/Compiling.html#spack-development explains why it is present, but `spack diy` was deprecated quite long ago.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]: Not sure if it is a good idea to introduce a new flag for old Kokkos version. Perhaps we can limit to 4.2: as we will not be able to test on all old versions.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]:
```suggestion
```

If the variant is conditionally defined.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py:390]: Are we sure that the compiler is always `amdclang++`?

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py:399]: Same here. Feel strange to hard code compiler name. I am not sure that it always works for external cuda install.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py]: Option name is quite long, should we try to make it shorter?

**Review by cedricchevalier19 — COMMENTED** [review]
I will need more time to check if it is working on non full spack software stack.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py]: when working with a list like this, I usually allow having a `when` portion that can be used to limit the variant.

Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Sure, we just have to pull it out of the list or make the list smarter. E.g. by adding an optional 3rd element or just directly make it a dict.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py:390]: `amdclang++` is the most likely case. It could potentially be just regular upstream clang if it was built with amdgpu target support. I believe this only becomes relevant once there is compiler-as-dependency support in Spack and if there is a way to define a `cuda_compiler` and `hip_compiler` somehow.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Shortened it.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py]: Removed.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py]: added condition

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]
2 failed specs: `kokkos~serial~shared`, `kokkos~serial~shared+cmake_lang` (resolved, was not legal)

**Comment by dalg24** [comment]
I don't know how Spack usually handles this and whether it needs to be caught before we attempt to configure Kokkos but explicitly disabling the serial backend without specifying what alternative host backend you want is not legal

**Comment by rbberger** [comment]
@bernhardkaindl as mentioned by @dalg24 , disabling `serial` without alternative makes no sense. I've now added a conflict that requires at least one active backend.

As for adopting `CachedCMakePackage`, that is a project decision that I will leave to the actual Kokkos package maintainers. I don't see any immediate downsides. If that is desired it should probably be its own PR.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [inline var/spack/repos/builtin/packages/kokkos/package.py:327]: You may also update the error message below.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (cedricchevalier19) [reply var/spack/repos/builtin/packages/kokkos/package.py:390]: Other occurrences of `CMAKE_HIP_COMPILER` use `clang++.` 

https://github.com/spack/spack/blob/9ac261af584c1ad6bdbcbad8b5ab1aec9283fa11/var/spack/repos/builtin/packages/py-onnxruntime/package.py#L201

https://github.com/spack/spack/blob/9ac261af584c1ad6bdbcbad8b5ab1aec9283fa11/lib/spack/spack/build_systems/cached_cmake.py#L285

**Review by cedricchevalier19 — APPROVED** [review]
Thank you, @rbberger; it is an essential addition to the Kokkos recipe.

As for the `CachedCMakePackage,` I agree that if we adopt it, a separate PR makes more sense.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py:390]: I've asked the AMD folks from the ElCapitan COE and they also say it should be `amdclang++`. @bvanessen @pranav-sivaraman do you know any particular reason why `clang++` was chosen in `CachedCMakePackage`?

Comment (pranav-sivaraman) [reply var/spack/repos/builtin/packages/kokkos/package.py:390]: I'm not sure why `clang++` was chosen, but I do think `amdclang++` should be the default compiler to match the rest of Spack CI until a virtual `hip_compiler` gets added.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos/package.py:327]: Updated.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (pranav-sivaraman) [reply var/spack/repos/builtin/packages/kokkos/package.py:390]: I guess for now I would avoid using `cached_cmake.py` since the way to deal with HIP compilers will change in the future. I'm not sure if there are many packages that are looking to compile with different HIP compilers so Kokkos should just implement it's own solution and it can be refactored some other day.

**Review by bernhardkaindl — APPROVED** [review]
Hi, to clean-up my review for this PR:
1. I've no specific knowledge about `kokkos`, so don't see this just as a message to the maintainers of this recipe.
2. I retract my change request from the failure (can only be done as an approval)
3. @rbberger / @cedricchevalier19 / @pranav-sivaraman:

I'd like to ask if you have further changes for this PR, or if merging should wait for some reason:

Please notify a `spack` member (such as me, or others) when this PR is approved by this recipe's maintainers and should be merged:

Preconditions, as I see it:
- No open change request (besides mine, none was marked as "changes requested")
- All conversations are formally declared as resolved:
  - @cedricchevalier19 I see you started a few conversations:
    - If you want to facilitate merging, please check if you see a button "Resolve conversation" below each of the conversations, and if you see it, it would be nice if you could confirm that the conversation is resolved by adding a comment to it that says so. Then please click the "Resolve conversation" button to resolve and hide the conversation (can be un-resolved later, if needed)

When all is fine for a merge (CI pipeline needs to pass - is running now) and everything is resolved, just mention e.g. me as @bernhardkaindl.

PS: You can formally declare your approval (or request changes) status on the Tab "Files changed" by opening "Review changes", by selecting "Approve" or "Request changes" and click on "Submit review" (that makes your status on the PR unambiguous)

PPS: This is no urgency, just a reminder to respond/approve/request changes as I saw further activity on this PR.

**Review by cedricchevalier19 — APPROVED** [review]

**Comment by cedricchevalier19** [comment]
@bernhardkaindl, I cannot mark the conversations as resolved; perhaps @rbberger can, but I do not have the button.
I would be happy if this PR could be merged soon.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/kokkos/package.py:390]: As I see the message "This comment was marked as resolved." and the comment by @cedricchevalier19, I mark this conversation as resolved.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/kokkos/package.py:399]: As I see the message "This comment was marked as resolved." and the comment by @cedricchevalier19, I mark this conversation as resolved.

File: var/spack/repos/builtin/packages/kokkos/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/kokkos/package.py]: As per the comment by @cedricchevalier19, and this is outdated, I mark this conversation as resolved.

**Comment by bernhardkaindl** [comment]
@cedricchevalier19, @rbberger:
> @bernhardkaindl, I cannot mark the conversations as resolved; perhaps @rbberger can, but I do not have the button.
> I would be happy if this PR could be merged soon.

Thanks, I resolved them now:
> All conversations are resolved
> 6 resolved conversations

As I now assume that everybody is fine with merging it, I'll enable auto-merge.

If anyone like to block the merge, you can try to request changes, or it may be enough to start a conversation in the review. Other than that, @cedricchevalier19 has a new PR https://github.com/spack/spack/pull/47330 on kokkos, that can be also used in case small cleanups shall be added, or a new PR can always be opened.

**Pull request merged**
